hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java

```diff
@@ -18,27 +18,220 @@
+import static org.apache.hadoop.hdds.scm.net.NetConstants.PATH_SEPARATOR_STR;
+import static org.apache.hadoop.ozone.OzoneConsts.OZONE_OFS_URI_SCHEME;
 
 /**
  * Executes deleteVolume call for the shell.
  */
 @Command(name = "delete",
-    description = "deletes a volume if it is empty")
+    description = "deletes a volume")
 public class DeleteVolumeHandler extends VolumeHandler {
+  @CommandLine.Option(
+      names = {"-skipTrash"},
+      description = "Delete volume without trash"
+  )
+  private boolean bSkipTrash = false;
+  @CommandLine.Option(
+      names = {"-r"},
+      description = "Delete volume recursively"
+  )
+  private boolean bRecursive = false;
+  @CommandLine.Option(
+      names = {"-id", "--om-service-id"},
+      description = "Ozone Manager Service ID"
+  )
+  private String omServiceId;
+  private ExecutorService executor;
+  private List<String> bucketIdList = new ArrayList<>();
+  private AtomicInteger cleanedBucketCounter =
+      new AtomicInteger();
+  private int totalBucketCount;
+  private OzoneVolume vol;
+  private AtomicInteger numberOfBucketsCleaned = new AtomicInteger(0);
+  private volatile Throwable exception;
+  private static final int MAX_KEY_DELETE_BATCH_SIZE = 1000;
 
   @Override
   protected void execute(OzoneClient client, OzoneAddress address)
       throws IOException {
 
     String volumeName = address.getVolumeName();
-
+    try {
+      if (bRecursive) {
+        if (!bSkipTrash) {
+          out().printf("Use -skipTrash for recursive volume delete%n");
+          return;
+        }
+        if (OmUtils.isServiceIdsDefined(getConf()) &&
+            Strings.isNullOrEmpty(omServiceId)) {
+          out().printf("OmServiceID not provided, provide using " +
+              "-id <OM_SERVICE_ID>%n");
+          return;
+        }
+        vol = client.getObjectStore().getVolume(volumeName);
+        deleteVolumeRecursive();
+      }
+    } catch (InterruptedException e) {
+      out().printf("Exception while deleting volume recursively%n");
+      return;
+    }
     client.getObjectStore().deleteVolume(volumeName);
     out().printf("Volume %s is deleted%n", volumeName);
   }
+
+  private void deleteVolumeRecursive()
+      throws InterruptedException {
+    // Get all the buckets for given volume
+    Iterator<? extends OzoneBucket> bucketIterator =
+        vol.listBuckets(null);
+
+    while (bucketIterator.hasNext()) {
+      OzoneBucket bucket = bucketIterator.next();
+      bucketIdList.add(bucket.getName());
+      totalBucketCount++;
+    }
+    doCleanBuckets();
+  }
+
+  /**
+   * Clean OBS bucket recursively.
+   *
+   * @param  bucket OzoneBucket
+   * @return boolean
+   */
+  private boolean cleanOBSBucket(OzoneBucket bucket) {
+    ArrayList<String> keys = new ArrayList<>();
+    try {
+      if (!bucket.isLink()) {
+        Iterator<? extends OzoneKey> iterator = bucket.listKeys(null);
+        while (iterator.hasNext()) {
```

Comment:
Same issue here, if new keys are written by another client concurrently , the iterator reference is older and doesn't have the new keys and bucket delete fails. Ideal way to solve this to take a bucket lock on the server , iterate through all keys and delete. If we are sure that the bucket is to be deleted , taking a bucket lock is ok as it doesn't take a toll on other ops as we know we wont be doing any ops on this bucket.